File: python-client/src/hammerspoon_macapi/namespaces/audio.py
"""Typed audio observation and output-control methods."""

from __future__ import annotations

from typing import TYPE_CHECKING

from ..models import AudioInfo

if TYPE_CHECKING:
    from ..client import MacAPI


class AudioClient:
    """Namespace wrapper for the ``audio.*`` RPC methods."""

    def __init__(self, api: MacAPI) -> None:
        self._api = api

    async def get(self) -> AudioInfo:
        """Return the current default input and output device information."""
        return await self._api.call("audio.get", result_type=AudioInfo)

    async def set_volume(self, volume: float) -> None:
        """Set output volume in the inclusive range 0 through 100."""
        if not 0 <= volume <= 100:
            raise ValueError("volume must be between 0 and 100")
        await self._api.call("audio.setVolume", {"volume": volume}, result_type=type(None))

    async def set_muted(self, muted: bool) -> None:
        """Set the default output device mute state."""
        await self._api.call("audio.setMuted", {"muted": muted}, result_type=type(None))

File: hammerspoon/macapi/audio.lua
--- Default audio device inspection and output-control RPC service.
local audio = require("hs.audiodevice")
local validators = require("macapi.validators")
local common = require("macapi.common")
local protocol = require("macapi.protocol")
local M = { methods = {} }

--- Return default input/output device details.
M.methods["audio.get"] = function()
    local output = audio.defaultOutputDevice()
    local input = audio.defaultInputDevice()
    local result = {
        output = output and { name = output:name(), volume = output:volume(), muted = output:muted() } or nil,
        input = input and { name = input:name() } or nil,
    }
    if not output and not input then return protocol.empty_object() end
    return result
end

--- Set default output volume in the inclusive range 0..100.
M.methods["audio.setVolume"] = function(params)
    local volume, error = validators.required_number(params, "volume")
    if not volume then return nil, error end
    if volume < 0 or volume > 100 then return common.error("INVALID_PARAMS", "volume must be between 0 and 100") end
    local output = audio.defaultOutputDevice()
    if not output then return common.error("FEATURE_DISABLED", "no default output device") end
    local current = output:volume()
    if current ~= nil and math.abs(current - volume) < 0.01 then return nil end
    output:setVolume(volume)
    return nil
end

--- Set default output mute state.
M.methods["audio.setMuted"] = function(params)
    local muted, error = validators.required_boolean(params, "muted")
    if muted == nil then return nil, error end
    local output = audio.defaultOutputDevice()
    if not output then return common.error("FEATURE_DISABLED", "no default output device") end
    if output:muted() == muted then return nil end
    output:setMuted(muted)
    return nil
end

return M
